### DISCUSSION TITLE
Making the order of recent buffers in consult-buffer local to each window

### DISCUSSION THREAD

**Discussion by egr95** [minad/consult#1141] [Ideas]
When I run consult-buffer, the first candidates displayed are the previous buffers I have visited, from most to least recent (excluding buffers that are visible in other windows). Is it possible to configure this ordering to behave locally to each window? 

To be more specific, say I have a vertical window split, and I am looking at buffer A on the left, and B on the right. Suppose I switch to buffer C on the left split, and then jump to the right with other-window and switch to buffer D. If I then jump back to the left and run consult-buffer, the first entry will be B, whereas I would like it to be A. This is useful because often one might use a specific window split for a specific type of buffer, and it is convenient to quickly cycle through the recently-visited buffers specifically in that window split. This is also how VScode buffer-switching works.

I don't understand the consult internals particularly well, but as far as I could tell this is not something that is directly available in the configuration options. Perhaps it would be a matter of adding an additional `consult--buffer-sort` function that uses some sort of `window-parameter`?

**Comment by minad** [comment]
Hi! This sounds like a good idea. One can use the `window-prev-buffers` or `window-next-buffers` function to obtain a list of relevant buffers. Right now there is no configuration option and no sort function which does this. One has to replace the `:items` function of `consult--buffer-query`. Assuming that we have a `consult--buffer-sort-window-local` function, then this `:items` function would do:

~~~ elisp
(lambda () 
  (consult--buffer-query :sort 'window-local ;; Note that window-local does not yet exist.
                         :as #'consult--buffer-pair))
~~~

**Reply by minad** [reply]
Another approach could be the following:

~~~ elisp
(setf (plist-get consult--source-buffer :items)
      (lambda ()
        (consult--buffer-query
         :sort 'visibility
         :as #'consult--buffer-pair
         :buffer-list (delete-dups (nconc (mapcar #'car (window-prev-buffers)) (buffer-list))))))
~~~

I use the list of `window-prev-buffers` and prioritize them. However I still nconc the `buffer-list` such that no buffers are lost. Does this achieve the behavior you are looking for? You can evaluate this code in the scratch buffer and see if it works.

**Reply by egr95** [reply]
Been using this and it works a charm, thank you @minad !

(FTR, for the exact VSCode behavior I use `:sort 'nil` instead, which makes sure to keep things in order of window-local recency even when a buffer is open in a different frame or window)

**Reply by minad** [reply]
Thanks for the feedback. I am gradually trying to improve the Consult defaults so maybe this will be added at some point.

**Comment by ksqsf** [comment]
Hi @minad I have a similar but slightly different problem: I always use C-x b to jump back to the previous buffer, but if it is shown in a different window, it will be lowered to the bottom. This gets annoying very soon if you view the same buffer in two windows.  

In this case, a window-local sorting would help. I searched and found this discussion, but unfortunately the solution from 2024 does not work anymore (the displayed buffer is still lowered). 

Would it be possible there is a simpler solution for two-window-same-buffer case, since it does not really require window-local sorting for the two-window-same-buffer case? Thanks!
